compiler/rustc_error_codes/src/error_codes/E0311.md

```diff
@@ -0,0 +1,69 @@
+This error occurs when there is an unsatisfied outlives bound on a generic
+type parameter or associated type.
+
+Erroneous code example:
+
+```compile_fail,E0311
+use std::borrow::BorrowMut;
+
+trait NestedBorrowMut<U, V> {
+    fn nested_borrow_mut(&mut self) -> &mut V;
+}
+
+impl<T, U, V> NestedBorrowMut<U, V> for T
+where
+    T: BorrowMut<U>,
+    U: BorrowMut<V>,
+{
+    fn nested_borrow_mut(&mut self) -> &mut V {
+        let u_ref = self.borrow_mut();
+        let v_ref = u_ref.borrow_mut();
+        v_ref
+    }
+}
+```
+
+Why doesn't this code compile? It helps to look at the lifetime bounds that
+the compiler is automatically adding ("Lifetime Ellision", Chapter 10.3 in the
+Rust book) to the `nested_borrow_mut()` and `borrow_mut()` functions. In both
+cases the input is a reference to `self`, so the compiler attempts to assign
+the same lifetime to the input and output.
+
+Looking specifically at `nested_borrow_mut()`, we see that there are three
+object references to keep track of, along with their associated lifetimes:
+- `self` (which is a `&mut T`)
+- `u_ref` (which is a `&mut U`)
+- `v_ref` (which is a `&mut V`)
+
+The `borrow_mut()` method implicitly requires that that the input and output
+have the same lifetime bounds. Thus the lines: `let u_ref = self.borrow_mut();`
+and `let v_ref = u_ref.borrow_mut();` in `nested_borrow_mut()` above imply that
+`u_ref` and `self` must share a lifetime bound, and also that `v_ref` and
+`u_ref` share a lifetime bound. The problem is that the function signature for
+`nested_borrow_mut()` only gives the compiler information about the lifetimes
+of `self` and `v_ref` -- nothing about `u_ref`.
```

Comment:
i feel like this explanation isn't quite right. `u_ref`, `self` and `v_ref` don't necessarily need the same lifetime, as we may reborrow `self` or `u_ref` for the `borrow_mut` call.

The reason for that error is:
- due to lifetime elision, we have to return a reference with some elided lifetime `'0` from outside of the function.
- because of the signature of `borrow_mut`, `u_ref` has to be valid for at least as long as `v_ref`, meaning that we have some `&'1 V` where `'1: '0` holds.
- this requires `V: '1` which is not implied by the signature (`'1` then ends up being equal to `'0` but that doesn't really matter)

I am not sure whether that's more helpful, but explaining the behavior the following (smaller) example might be easier
```rust
fn no_restriction<T>(x: &()) -> &() {
    with_restriction::<T>(x)
}

fn with_restriction<'a, T: 'a>(x: &'a ()) -> &'a () {
    x
}
```

Reply:
Thanks for explaining this. I like your minimum working example much better: it makes it more clear exactly what the problem is.  I've replace my original example with your suggestion.

One thing that I missed the first few times that I read this is that the above code works fine if the generic parameter `T` is omitted. I added an example to that effect in the updated documentation because I figure that it might help make it more clear why the lifetime elision fails in this case.